modules/cli/src/main/scala/scala/cli/commands/DependencyUpdate.scala

```diff
@@ -0,0 +1,121 @@
+    CurrentParams.verbosity = verbosity
+
+    val inputs       = options.shared.inputsOrExit(args)
+    val logger       = options.shared.logger
+    val buildOptions = options.shared.buildOptions()
+
+    val crossSources =
+      CrossSources.forInputs(
+        inputs,
+        Sources.defaultPreprocessors(
+          buildOptions.scriptOptions.codeWrapper.getOrElse(CustomCodeWrapper),
+          buildOptions.archiveCache,
+          buildOptions.internal.javaClassNameVersionOpt
+        ),
+        logger
+      ).orExit(logger)
+
+    val scopedSources = crossSources.scopedSources(buildOptions).orExit(logger)
+
+    def generateActionableUpdateDiagnostic(scope: Scope)
+      : Seq[ActionableDependencyUpdateDiagnostic] = {
+      val sources = scopedSources.sources(scope, crossSources.sharedOptions(buildOptions))
+
+      if (verbosity >= 3)
+        pprint.err.log(sources)
+
+      val options = buildOptions.orElse(sources.buildOptions)
+      val actionableDiagnostics =
+        ActionablePreprocessor.generateActionableDiagnostics(options).orExit(logger)
+
+      actionableDiagnostics.collect {
+        case ad: ActionableDependencyUpdateDiagnostic => ad
+      }
+    }
+
+    val actionableMainUpdateDiagnostics = generateActionableUpdateDiagnostic(Scope.Main)
+    val actionableTestUpdateDiagnostics = generateActionableUpdateDiagnostic(Scope.Test)
+    val actionableUpdateDiagnostics =
+      (actionableMainUpdateDiagnostics ++ actionableTestUpdateDiagnostics).distinct
+
+    if (options.all)
+      updateDependencies(actionableUpdateDiagnostics, logger)
+  }
+
+  private def updateDependencies(
+    actionableUpdateDiagnostics: Seq[ActionableDependencyUpdateDiagnostic],
+    logger: Logger
+  ): Unit = {
+    actionableUpdateDiagnostics.foreach { diagnostic =>
+      diagnostic.positions
+        .collect { case file: Position.File => file }
+        .foreach {
+          case Position.File(Right(filePath), startPos, _) =>
+            val lineIndex = startPos._1
+            val appliedDiagnostic = splitFileContentByLine(filePath).map {
+              case (line, index) if index == lineIndex + 1 =>
+                updateDependency(line, startPos, diagnostic)
+              case (line, _) => line
+            }.mkString
+            os.write.over(filePath, appliedDiagnostic)
+            logger.message(s"Updated dependency to: ${diagnostic.to}")
+          case Position.File(Left(file), _, _) =>
+            logger.message(
+              s"scala-cli can't update dependency to:${diagnostic.to} in virtual file: $file"
+            )
+        }
+    }
+  }
+
+  private def splitFileContentByLine(file: os.Path) = {
+    val content       = os.read(file)
+    val startIndicies = Position.Raw.lineStartIndices(content).toList
+
+    def lsplit(pos: List[Int], s: String): List[String] =
+      pos match {
+        case x :: rest => s.substring(0, x) :: lsplit(rest.map(_ - x), s.substring(x))
```

Comment:
This is not tail-recursive, it seems this can blow the stack for large files. But also, as we typically only edit a handful of lines, maybe we shouldn't split the whole file like this… Better keep its content (as a large string), then change a line in it using start end end indices.